decompiler/pipeline/expressions/bitfieldcomparisonunrolling.py

```diff
@@ -0,0 +1,150 @@
+from logging import debug, warning
+from typing import Any, List, Optional, Tuple
+
+from decompiler.pipeline.stage import PipelineStage
+from decompiler.structures.graphs.basicblock import BasicBlock
+from decompiler.structures.graphs.branches import FalseCase, TrueCase, UnconditionalEdge
+from decompiler.structures.graphs.cfg import ControlFlowGraph
+from decompiler.structures.pseudo import Constant, Expression
+from decompiler.structures.pseudo.expressions import Variable
+from decompiler.structures.pseudo.instructions import Branch
+from decompiler.structures.pseudo.operations import BinaryOperation, Condition, OperationType, UnaryOperation
+from decompiler.task import DecompilerTask
+
+
+class BitFieldComparisonUnrolling(PipelineStage):
+    """
+    Transform bit-field compiler optimization to readable comparison:
+
+    var = 1 << amount;
+    if ((var & 0b11010) != 0) { ... }
+
+    // becomes:
+
+    if ( amount == 1 || amount == 3 || amount == 4 ) { ... }
+
+    This can subsequently be used to reconstruct switch-case statements.
+
+    This stage requires expression-propagation PipelineStage, such that bit-shift
+    gets forwarded into Branch.condition:
+
+    if ( (1 << amount) & bit_mask) == 0) ) { ... }
+    """
+
+    name = "bit-field-comparison-unrolling"
+    dependencies = ["expression-propagation"]
+
+    def run(self, task: DecompilerTask):
+        """Run the pipeline stage: Check all viable Branch-instructions."""
+        worklist = []
+        for block in task.graph:
+            if (matched_instruction := self._unfold_cases(block)) is not None:
+                switch_var, cases, is_negated = matched_instruction
+                worklist.append((block, switch_var, cases, is_negated))
+        for todo in worklist:
+            self._modify_cfg(task.graph, *todo)
+
+    def _modify_cfg(self, cfg: ControlFlowGraph, block: BasicBlock, switch_var: Variable, cases: List[int], is_negated_condition: bool):
+        """
+        Create nested if blocks for each case in unfolded values.
+        Note: with the Branch condition encountered so far (== 0x0), the node of the collected cases is adjacent to the FalseCase edge.
+        However, negated conditions may exist. In this case, pass condition type (flag) and swap successor nodes accordingly.
+        """
+        debug("modifying cfg")
+        true_edge, false_edge = self._tf_edges_from_block(cfg, block)
+        other_node, case_node = true_edge.sink, false_edge.sink
+        if is_negated_condition:
+            other_node, case_node = case_node, other_node
```

Comment:
I am a bit confused. You have the condition `1 << var & 0b11010) != 0`. Then `is_negated_condition` is True. But, this condition is the same as `var == 1 || var == 3 || var == 4`, so the `case_node` is the true-branch. This is a bit confusing.

Reply:
It is indeed confusing... In the case of `1 << var & 0b11010) != 0` the case node is incident to the true branch. But we mostly have `1 << var & 0b11010) == 0` and the case node is at the false edge. I try to clarify this.

Reply:
I made it more explicit which edge type we expect to go to the case node.